// @revideo/ffmpeg/dist/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ffmpegSettings = void 0;
const pathToFfmpeg = require("ffmpeg-static");
const pathToFfprobe = require("ffprobe-static");
const ffmpegLogLevels = [
    'quiet',
    'panic',
    'fatal',
    'error',
    'warning',
    'info',
    'verbose',
    'debug',
    'trace',
];
class FfmpegSettingState {
    constructor() {
        this.ffmpegPath = pathToFfmpeg;
        this.ffprobePath = pathToFfprobe.path;
        // Use the FFMPEG_PATH environment variable if it is set
        if (process.env.FFMPEG_PATH) {
            this.ffmpegPath = process.env.FFMPEG_PATH;
        }
        // Use the FFPROBE_PATH environment variable if it is set
        if (process.env.FFPROBE_PATH) {
            this.ffprobePath = process.env.FFPROBE_PATH;
        }
        this.logLevel = 'error';
        // Use the FFMPEG_LOG_LEVEL environment variable if it is set
        if (process.env.FFMPEG_LOG_LEVEL &&
            ffmpegLogLevels.includes(process.env.FFMPEG_LOG_LEVEL)) {
            this.logLevel = process.env.FFMPEG_LOG_LEVEL;
        }
    }
    getFfmpegPath() {
        return this.ffmpegPath;
    }
    setFfmpegPath(ffmpegPath) {
        this.ffmpegPath = ffmpegPath;
    }
    getFfprobePath() {
        return this.ffprobePath;
    }
    setFfprobePath(ffprobePath) {
        this.ffprobePath = ffprobePath;
    }
    getLogLevel() {
        return this.logLevel;
    }
    setLogLevel(logLevel) {
        this.logLevel = logLevel;
    }
}
exports.ffmpegSettings = new FfmpegSettingState();
