// @revideo/core/lib/scenes/Variables.js
import { createSignal } from '../signals';
export class Variables {
    constructor(scene) {
        this.scene = scene;
        this.signals = {};
        this.variables = {};
        /**
         * Reset all stored signals.
         */
        this.handleReset = () => {
            this.signals = {};
        };
        scene.onReset.subscribe(this.handleReset);
    }
    /**
     * Get variable signal if exists or create signal if not
     *
     * @param name - The name of the variable.
     * @param initial - The initial value of the variable. It will be used if the
     *                  variable was not configured from the outside.
     */
    get(name, initial) {
        var _a;
        (_a = this.signals)[name] ?? (_a[name] = createSignal(this.variables[name] ?? initial));
        return () => this.signals[name]();
    }
    /**
     * Update all signals with new project variable values.
     */
    updateSignals(variables) {
        this.variables = variables;
        Object.keys(variables).map(variableName => {
            if (variableName in this.signals) {
                this.signals[variableName](variables[variableName]);
            }
        });
    }
}
