// @revideo/2d/lib/code/LezerHighlighter.js
import { highlightTree } from '@lezer/highlight';
import { useLogger } from '@revideo/core';
import { defaultTokenize } from './CodeTokenizer';
export class LezerHighlighter {
    static registerParser(parser, dialect = '') {
        this.parserMap.set(dialect, parser);
    }
    constructor(style) {
        this.style = style;
        this.classLookup = new Map();
        for (const rule of this.style.module?.getRules().split('\n') ?? []) {
            const match = rule.match(LezerHighlighter.classRegex);
            if (!match) {
                continue;
            }
            const className = match[1];
            const color = match[2].trim();
            this.classLookup.set(className, color);
        }
    }
    initialize() {
        return true;
    }
    prepare(code, dialect) {
        const parser = LezerHighlighter.parserMap.get(dialect);
        if (!parser) {
            if (dialect !== '') {
                useLogger().warn(`No parser found for dialect: ${dialect}`);
            }
            return null;
        }
        const colorLookup = new Map();
        const tree = parser.parse(code);
        highlightTree(tree, this.style, (from, to, classes) => {
            const color = this.classLookup.get(classes);
            if (!color) {
                return;
            }
            const cursor = tree.cursorAt(from, 1);
            do {
                const id = this.getNodeId(cursor.node);
                colorLookup.set(id, color);
            } while (cursor.next() && cursor.to <= to);
        });
        return {
            tree,
            code,
            colorLookup,
        };
    }
    highlight(index, cache) {
        if (!cache) {
            return {
                color: null,
                skipAhead: 0,
            };
        }
        const node = cache.tree.resolveInner(index, 1);
        const id = this.getNodeId(node);
        const color = cache.colorLookup.get(id);
        if (color) {
            return {
                color,
                skipAhead: node.to - index,
            };
        }
        let skipAhead = 0;
        if (!node.firstChild) {
            skipAhead = node.to - index;
        }
        return {
            color: null,
            skipAhead,
        };
    }
    tokenize(code, dialect) {
        const parser = LezerHighlighter.parserMap.get(dialect);
        if (!parser) {
            if (dialect !== '') {
                useLogger().warn(`No parser found for dialect: ${dialect}`);
            }
            return defaultTokenize(code);
        }
        const tree = parser.parse(code);
        const cursor = tree.cursor();
        const tokens = [];
        let current = 0;
        do {
            if (!cursor.node.firstChild) {
                if (cursor.from > current) {
                    tokens.push(code.slice(current, cursor.from));
                }
                if (cursor.from < cursor.to) {
                    tokens.push(code.slice(cursor.from, cursor.to));
                }
                current = cursor.to;
            }
        } while (cursor.next());
        return tokens;
    }
    getNodeId(node) {
        if (!node.parent) {
            return -1;
        }
        // NOTE: They don't want us to know about this property.
        // We need a way to persistently identify nodes and this seems to work.
        // Perhaps it could break if the tree is edited? But we don't do that. Yet.
        return node.index;
    }
}
LezerHighlighter.parserMap = new Map();
LezerHighlighter.classRegex = /\.(\S+).*color:([^;]+)/;
