// @revideo/renderer/lib/client/render.js
import { Renderer, Vector2 } from '@revideo/core';
/**
 * Render the project.
 * @param project - The project to render.
 * @param range - The range of frames to render.
 */
export const render = async (project, workerId, totalNumOfWorkers, startInSeconds, endInSeconds, hiddenFolderId, videoWidth, videoHeight) => {
    try {
        const renderer = new Renderer(project);
        const { firstGlobalFrame, lastGlobalFrame } = await getGlobalFirstAndLastFrame(project, renderer, startInSeconds, endInSeconds);
        const { firstWorkerFrame, lastWorkerFrame } = await getWorkerFirstAndLastFrame(firstGlobalFrame, lastGlobalFrame, workerId, totalNumOfWorkers);
        const renderSettings = {
            ...project.meta.getFullRenderingSettings(),
            name: project.name,
            hiddenFolderId: hiddenFolderId,
            range: [
                renderer.frameToTime(firstWorkerFrame),
                renderer.frameToTime(lastWorkerFrame),
            ],
        };
        if (videoWidth && videoHeight) {
            renderSettings.size = new Vector2({ x: videoWidth, y: videoHeight });
        }
        await renderer.render(renderSettings);
        window.onRenderComplete();
    }
    catch (e) {
        window.onRenderFailed(e.message);
    }
};
/**
 * Calculate the first and last "global" frame, i.e. frame independent of worker
 */
async function getGlobalFirstAndLastFrame(project, renderer, startSecondFromUser, endSecondFromUser) {
    const firstGlobalFrame = renderer.timeToFrame(startSecondFromUser);
    let lastGlobalFrame;
    // If end value is provided by user, lastFrame can be obtained.
    if (isFinite(endSecondFromUser)) {
        lastGlobalFrame = renderer.timeToFrame(endSecondFromUser);
    }
    else {
        // Otherwise, endSecondFromUser is infinity, so lastGlobalFrame is the duration of the video.
        const settings = {
            ...project.meta.getFullRenderingSettings(),
            name: project.name,
        };
        lastGlobalFrame = await renderer.getNumberOfFrames(settings);
    }
    return { firstGlobalFrame, lastGlobalFrame };
}
async function getWorkerFirstAndLastFrame(firstGlobalFrame, lastGlobalFrame, workerId, totalNumOfWorkers) {
    const videoDurationInFrames = lastGlobalFrame - firstGlobalFrame + 1;
    const framesPerWorker = Math.ceil(videoDurationInFrames / totalNumOfWorkers);
    const firstWorkerFrame = firstGlobalFrame + framesPerWorker * workerId;
    const lastWorkerFrame = Math.min(firstWorkerFrame + framesPerWorker, lastGlobalFrame);
    return { firstWorkerFrame, lastWorkerFrame };
}
