// @revideo/core/lib/signals/CompoundSignalContext.js
import { map } from '../tweening';
import { SignalContext } from './SignalContext';
import { isReactive, modify } from './utils';
export class CompoundSignalContext extends SignalContext {
    constructor(entries, parser, initial, interpolation, owner = undefined, extensions = {}) {
        var _a;
        super(undefined, interpolation, owner, parser, extensions);
        this.entries = entries;
        this.signals = [];
        this.parser = parser;
        for (const entry of entries) {
            let key;
            let signal;
            if (Array.isArray(entry)) {
                [key, signal] = entry;
                (_a = signal.context).owner ?? (_a.owner = this);
            }
            else {
                key = entry;
                signal = new SignalContext(modify(initial, value => parser(value)[entry]), map, owner ?? this.invokable).toSignal();
            }
            this.signals.push([key, signal]);
            Object.defineProperty(this.invokable, key, { value: signal });
        }
    }
    toSignal() {
        return this.invokable;
    }
    parse(value) {
        return this.parser(value);
    }
    getter() {
        return this.parse((Object.fromEntries(this.signals.map(([key, property]) => [key, property()]))));
    }
    setter(value) {
        if (isReactive(value)) {
            for (const [key, property] of this.signals) {
                property(() => this.parser(value())[key]);
            }
        }
        else {
            const parsed = this.parse(value);
            for (const [key, property] of this.signals) {
                property(parsed[key]);
            }
        }
        return this.owner;
    }
    reset() {
        for (const [, signal] of this.signals) {
            signal.reset();
        }
        return this.owner;
    }
    save() {
        for (const [, signal] of this.signals) {
            signal.save();
        }
        return this.owner;
    }
    isInitial() {
        for (const [, signal] of this.signals) {
            if (!signal.isInitial()) {
                return false;
            }
        }
        return true;
    }
    raw() {
        return Object.fromEntries(this.signals.map(([key, property]) => [key, property.context.raw()]));
    }
}
