// @revideo/2d/lib/curves/CubicBezierSegment.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Vector2, lazy } from '@revideo/core';
import { bezierCurveTo } from '../utils';
import { Polynomial2D } from './Polynomial2D';
import { PolynomialSegment } from './PolynomialSegment';
/**
 * A spline segment representing a cubic Bézier curve.
 */
export class CubicBezierSegment extends PolynomialSegment {
    get points() {
        return [this.p0, this.p1, this.p2, this.p3];
    }
    constructor(p0, p1, p2, p3) {
        super(new Polynomial2D(p0, 
        // 3*(-p0+p1)
        p0.flipped.add(p1).scale(3), 
        // 3*p0-6*p1+3*p2
        p0.scale(3).sub(p1.scale(6)).add(p2.scale(3)), 
        // -p0+3*p1-3*p2+p3
        p0.flipped.add(p1.scale(3)).sub(p2.scale(3)).add(p3)), CubicBezierSegment.getLength(p0, p1, p2, p3));
        this.p0 = p0;
        this.p1 = p1;
        this.p2 = p2;
        this.p3 = p3;
    }
    split(t) {
        const a = new Vector2(this.p0.x + (this.p1.x - this.p0.x) * t, this.p0.y + (this.p1.y - this.p0.y) * t);
        const b = new Vector2(this.p1.x + (this.p2.x - this.p1.x) * t, this.p1.y + (this.p2.y - this.p1.y) * t);
        const c = new Vector2(this.p2.x + (this.p3.x - this.p2.x) * t, this.p2.y + (this.p3.y - this.p2.y) * t);
        const d = new Vector2(a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t);
        const e = new Vector2(b.x + (c.x - b.x) * t, b.y + (c.y - b.y) * t);
        const p = new Vector2(d.x + (e.x - d.x) * t, d.y + (e.y - d.y) * t);
        const left = new CubicBezierSegment(this.p0, a, d, p);
        const right = new CubicBezierSegment(p, e, c, this.p3);
        return [left, right];
    }
    doDraw(context) {
        bezierCurveTo(context, this.p1, this.p2, this.p3);
    }
    static getLength(p0, p1, p2, p3) {
        // Let the browser do the work for us instead of calculating the arclength
        // manually.
        CubicBezierSegment.el.setAttribute('d', `M ${p0.x} ${p0.y} C ${p1.x} ${p1.y} ${p2.x} ${p2.y} ${p3.x} ${p3.y}`);
        return CubicBezierSegment.el.getTotalLength();
    }
}
__decorate([
    lazy(() => document.createElementNS('http://www.w3.org/2000/svg', 'path'))
], CubicBezierSegment, "el", void 0);
