// @revideo/2d/lib/curves/LineSegment.js
import { lineTo, moveTo } from '../utils';
import { Segment } from './Segment';
export class LineSegment extends Segment {
    constructor(from, to) {
        super();
        this.from = from;
        this.to = to;
        this.vector = to.sub(from);
        this.length = this.vector.magnitude;
        this.normal = this.vector.perpendicular.normalized.safe;
        this.points = [from, to];
    }
    get arcLength() {
        return this.length;
    }
    draw(context, start = 0, end = 1, move = false) {
        const from = this.from.add(this.vector.scale(start));
        const to = this.from.add(this.vector.scale(end));
        if (move) {
            moveTo(context, from);
        }
        lineTo(context, to);
        return [
            {
                position: from,
                tangent: this.normal.flipped,
                normal: this.normal,
            },
            {
                position: to,
                tangent: this.normal,
                normal: this.normal,
            },
        ];
    }
    getPoint(distance) {
        const point = this.from.add(this.vector.scale(distance));
        return {
            position: point,
            tangent: this.normal.flipped,
            normal: this.normal,
        };
    }
}
