// @revideo/core/lib/meta/MetaFile.js
var _a;
import { Semaphore, useLogger } from '../utils';
/**
 * Represents the meta file of a given entity.
 *
 * @remarks
 * This class is used exclusively by our Vite plugin as a bridge between
 * physical files and their runtime representation.
 *
 * @typeParam T - The type of the data stored in the meta file.
 *
 * @internal
 */
export class MetaFile {
    constructor(name, source = false) {
        this.name = name;
        this.source = source;
        this.lock = new Semaphore();
        this.ignoreChange = false;
        this.cache = null;
        this.metaField = null;
        this.handleChanged = async () => {
            if (import.meta.hot && this.metaField && !this.ignoreChange) {
                const data = this.metaField.serialize();
                await this.lock.acquire();
                try {
                    // TODO Consider debouncing saving the meta file.
                    await this.saveData(data);
                }
                catch (e) {
                    useLogger().error(e);
                }
                this.lock.release();
            }
        };
    }
    attach(field) {
        if (this.metaField)
            return;
        this.metaField = field;
        if (this.cache) {
            this.metaField.set(this.cache);
        }
        this.metaField?.onChanged.subscribe(this.handleChanged);
    }
    async saveData(data) {
        if (this.source === false) {
            return;
        }
        if (!this.source) {
            throw new Error(`The meta file for ${this.name} is missing.`);
        }
        if (_a.sourceLookup[this.source]) {
            throw new Error(`Metadata for ${this.name} is already being updated`);
        }
        const source = this.source;
        await new Promise((resolve, reject) => {
            setTimeout(() => {
                delete _a.sourceLookup[source];
                reject(`Connection timeout when updating metadata for ${this.name}`);
            }, 1000);
            _a.sourceLookup[source] = () => {
                delete _a.sourceLookup[source];
                resolve();
            };
            import.meta.hot.send('revideo:meta', {
                source,
                data,
            });
        });
    }
    /**
     * Load new metadata from a file.
     *
     * @remarks
     * This method is called during hot module replacement.
     *
     * @param data - New metadata.
     */
    loadData(data) {
        this.ignoreChange = true;
        this.cache = data;
        this.metaField?.set(data);
        this.ignoreChange = false;
    }
}
_a = MetaFile;
MetaFile.sourceLookup = {};
(() => {
    if (import.meta.hot) {
        import.meta.hot.on('revideo:meta-ack', ({ source }) => {
            _a.sourceLookup[source]?.();
        });
    }
})();
