// @revideo/core/lib/meta/MetaField.js
import { ValueDispatcher } from '../events';
/**
 * Represents an entry in the meta file.
 *
 * @typeParam TSerializedValue - The type used to store this field in the meta
 *                               file.
 * @typeParam TValue - The runtime type of this field.
 */
export class MetaField {
    /**
     * Triggered when the data of this field changes.
     *
     * @eventProperty
     */
    get onChanged() {
        return this.value.subscribable;
    }
    /**
     * Triggered when the field becomes disabled or enabled.
     *
     * @eventProperty
     */
    get onDisabled() {
        return this.disabled.subscribable;
    }
    /**
     * @param name - The name of this field displayed in the editor.
     * @param initial - The initial value of this field.
     */
    constructor(name, initial) {
        this.name = name;
        this.initial = initial;
        /**
         * The type of this field used by the editor to display the correct input.
         */
        this.type = undefined;
        this.spacing = false;
        this.description = '';
        this.disabled = new ValueDispatcher(false);
        this.value = new ValueDispatcher(initial);
    }
    /**
     * Get the current value.
     */
    get() {
        return this.value.current;
    }
    /**
     * Set the current value.
     *
     * @param value - The new value.
     */
    set(value) {
        this.value.current = this.parse(value);
    }
    /**
     * Convert a serialized value into a runtime type.
     *
     * @param value - The serialized value.
     */
    parse(value) {
        return value;
    }
    /**
     * Serialize the value of this field.
     */
    serialize() {
        return this.value.current;
    }
    /**
     * Create a clone of this field.
     */
    clone() {
        return new this.constructor(this.name, this.get());
    }
    /**
     * Disable or enable the field in the editor.
     *
     * @param value - Whether the field should be disabled.
     */
    disable(value = true) {
        this.disabled.current = value;
        return this;
    }
    /**
     * Add or remove spacing at the beginning of this field.
     *
     * @param value - Whether to include the spacing.
     */
    space(value = true) {
        this.spacing = value;
        return this;
    }
    /**
     * Set the description of this field.
     *
     * @param description - The description.
     */
    describe(description) {
        this.description = description;
        return this;
    }
}
