// @revideo/core/lib/app/Logger.js
import { EventDispatcher } from '../events';
export var LogLevel;
(function (LogLevel) {
    LogLevel["Error"] = "error";
    LogLevel["Warn"] = "warn";
    LogLevel["Info"] = "info";
    LogLevel["Http"] = "http";
    LogLevel["Verbose"] = "verbose";
    LogLevel["Debug"] = "debug";
    LogLevel["Silly"] = "silly";
})(LogLevel || (LogLevel = {}));
export class Logger {
    constructor() {
        this.logged = new EventDispatcher();
        this.history = [];
        this.profilers = {};
    }
    /**
     * Triggered when a new message is logged.
     */
    get onLogged() {
        return this.logged.subscribable;
    }
    log(payload) {
        this.logged.dispatch(payload);
        this.history.push(payload);
    }
    error(payload) {
        this.logLevel(LogLevel.Error, payload);
    }
    warn(payload) {
        this.logLevel(LogLevel.Warn, payload);
    }
    info(payload) {
        this.logLevel(LogLevel.Info, payload);
    }
    http(payload) {
        this.logLevel(LogLevel.Http, payload);
    }
    verbose(payload) {
        this.logLevel(LogLevel.Verbose, payload);
    }
    debug(payload) {
        this.logLevel(LogLevel.Debug, payload);
    }
    silly(payload) {
        this.logLevel(LogLevel.Silly, payload);
    }
    logLevel(level, payload) {
        const result = typeof payload === 'string' ? { message: payload } : payload;
        result.level = level;
        this.log(result);
    }
    profile(id, payload) {
        const time = performance.now();
        if (this.profilers[id]) {
            const timeEnd = this.profilers[id];
            delete this.profilers[id];
            const result = payload ?? { message: id };
            result.level ?? (result.level = LogLevel.Debug);
            result.durationMs = time - timeEnd;
            this.log(result);
            return;
        }
        this.profilers[id] = time;
    }
}
