// @revideo/core/lib/app/PlaybackStatus.js
/**
 * A read-only representation of the playback.
 */
export class PlaybackStatus {
    constructor(playback) {
        this.playback = playback;
    }
    /**
     * Convert seconds to frames using the current framerate.
     *
     * @param seconds - The seconds to convert.
     */
    secondsToFrames(seconds) {
        return Math.ceil(seconds * this.playback.fps);
    }
    /**
     * Convert frames to seconds using the current framerate.
     *
     * @param frames - The frames to convert.
     */
    framesToSeconds(frames) {
        return frames / this.playback.fps;
    }
    get time() {
        return this.framesToSeconds(this.playback.frame);
    }
    get frame() {
        return this.playback.frame;
    }
    get speed() {
        return this.playback.speed;
    }
    get fps() {
        return this.playback.fps;
    }
    get state() {
        return this.playback.state;
    }
    /**
     * The time passed since the last frame in seconds.
     */
    get deltaTime() {
        return this.framesToSeconds(1) * this.speed;
    }
}
