// @revideo/core/lib/app/ProjectMetadata.js
import { ColorMetaField, EnumMetaField, ExporterMetaField, MetaField, NumberMetaField, ObjectMetaField, RangeMetaField, Vector2MetaField, } from '../meta';
import { Vector2 } from '../types';
import { ColorSpaces, FrameRates, Scales } from './presets';
function createProjectMetadata(project) {
    const meta = {
        version: new MetaField('version', 1),
        shared: new ObjectMetaField('General', {
            background: new ColorMetaField('background', null),
            range: new RangeMetaField('range', [0, Infinity]),
            size: new Vector2MetaField('resolution', new Vector2(1920, 1080)),
            audioOffset: new NumberMetaField('audio offset', 0),
        }),
        preview: new ObjectMetaField('Preview', {
            fps: new NumberMetaField('frame rate', 30)
                .setPresets(FrameRates)
                .setRange(1),
            resolutionScale: new EnumMetaField('scale', Scales, 1),
        }),
        rendering: new ObjectMetaField('Rendering', {
            fps: new NumberMetaField('frame rate', 60)
                .setPresets(FrameRates)
                .setRange(1),
            resolutionScale: new EnumMetaField('scale', Scales, 1),
            colorSpace: new EnumMetaField('color space', ColorSpaces),
            exporter: new ExporterMetaField('exporter', project),
        }),
    };
    meta.shared.audioOffset.disable(!project.audio);
    return meta;
}
export class ProjectMetadata extends ObjectMetaField {
    constructor(project) {
        super('project', createProjectMetadata(project));
    }
    getFullPreviewSettings() {
        return {
            ...this.shared.get(),
            ...this.preview.get(),
        };
    }
    getFullRenderingSettings() {
        return {
            ...this.shared.get(),
            ...this.rendering.get(),
        };
    }
}
