// @revideo/core/lib/events/EventDispatcherBase.js
/**
 * A base for dispatching {@link Subscribable}s.
 *
 * @typeParam TValue - The type of the argument passed to subscribers.
 * @typeParam THandler - The type of the callback function.
 */
export class EventDispatcherBase {
    constructor() {
        this.subscribable = new Subscribable(this);
        this.subscribers = new Set();
    }
    /**
     * {@inheritDoc Subscribable.subscribe}
     */
    subscribe(handler) {
        this.subscribers.add(handler);
        return () => this.unsubscribe(handler);
    }
    /**
     * {@inheritDoc Subscribable.unsubscribe}
     */
    unsubscribe(handler) {
        this.subscribers.delete(handler);
    }
    /**
     * Unsubscribe all subscribers from the event.
     */
    clear() {
        this.subscribers.clear();
    }
    notifySubscribers(value) {
        return [...this.subscribers].map(handler => handler(value));
    }
}
/**
 * Provides safe access to the public interface of {@link EventDispatcherBase}.
 *
 * @remarks
 * External classes can use it to subscribe to an event without being able to
 * dispatch it.
 *
 * @typeParam TValue - The type of the argument passed to subscribers.
 * @typeParam THandler - The type of the callback function.
 */
export class Subscribable {
    constructor(dispatcher) {
        this.dispatcher = dispatcher;
    }
    /**
     * Subscribe to the event.
     *
     * @param handler - The handler to invoke when the event occurs.
     *
     * @returns A callback function that cancels the subscription.
     */
    subscribe(handler) {
        return this.dispatcher.subscribe(handler);
    }
    /**
     * Unsubscribe from the event.
     *
     * @param handler - The handler to unsubscribe.
     */
    unsubscribe(handler) {
        this.dispatcher.unsubscribe(handler);
    }
}
