// @revideo/2d/lib/components/Audio.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Audio_1;
import { DependencyContext, PlaybackState, viaProxy } from '@revideo/core';
import { computed, nodeName } from '../decorators';
import { Media } from './Media';
let Audio = Audio_1 = class Audio extends Media {
    constructor(props) {
        super(props);
    }
    mediaElement() {
        return this.audio();
    }
    seekedMedia() {
        return this.seekedAudio();
    }
    fastSeekedMedia() {
        return this.fastSeekedAudio();
    }
    audio() {
        const src = viaProxy(this.fullSource());
        const key = `${this.key}/${src}`;
        let audio = Audio_1.pool[key];
        if (!audio) {
            audio = document.createElement('audio');
            audio.crossOrigin = 'anonymous';
            audio.src = src;
            Audio_1.pool[key] = audio;
        }
        if (audio.readyState < 2) {
            DependencyContext.collectPromise(new Promise(resolve => {
                const onCanPlay = () => {
                    resolve();
                    audio.removeEventListener('canplay', onCanPlay);
                };
                const onError = () => {
                    const reason = this.getErrorReason(audio.error?.code);
                    console.log(`ERROR: Error loading audio: ${src}, ${reason}`);
                };
                audio.addEventListener('canplay', onCanPlay);
                audio.addEventListener('error', onError);
            }));
        }
        return audio;
    }
    seekedAudio() {
        const audio = this.audio();
        audio.addEventListener('ended', () => {
            this.pause();
        });
        if (!(this.time() < audio.duration)) {
            this.pause();
            return audio;
        }
        const time = this.clampTime(this.time());
        audio.playbackRate = this.playbackRate();
        if (!audio.paused) {
            audio.pause();
        }
        if (this.lastTime === time) {
            return audio;
        }
        this.setCurrentTime(time);
        return audio;
    }
    fastSeekedAudio() {
        const audio = this.audio();
        if (!(this.time() < audio.duration)) {
            this.pause();
            return audio;
        }
        const time = this.clampTime(this.time());
        audio.playbackRate = this.playbackRate();
        if (this.lastTime === time) {
            return audio;
        }
        const playing = this.playing() && time < audio.duration && audio.playbackRate > 0;
        if (playing) {
            if (audio.paused) {
                DependencyContext.collectPromise(audio.play());
            }
        }
        else {
            if (!audio.paused) {
                audio.pause();
            }
        }
        if (Math.abs(audio.currentTime - time) > 0.3) {
            this.setCurrentTime(time);
        }
        else if (!playing) {
            audio.currentTime = time;
        }
        return audio;
    }
    async draw(context) {
        const playbackState = this.view().playbackState();
        playbackState === PlaybackState.Playing ||
            playbackState === PlaybackState.Presenting
            ? this.fastSeekedAudio()
            : this.seekedAudio();
        context.save();
        context.restore();
        await this.drawChildren(context);
    }
};
Audio.pool = {};
__decorate([
    computed()
], Audio.prototype, "audio", null);
__decorate([
    computed()
], Audio.prototype, "seekedAudio", null);
__decorate([
    computed()
], Audio.prototype, "fastSeekedAudio", null);
Audio = Audio_1 = __decorate([
    nodeName('Audio')
], Audio);
export { Audio };
