// @revideo/2d/lib/components/Txt.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Txt_1;
import { DEFAULT, all, capitalize, threadable, } from '@revideo/core';
import { computed, initial, nodeName, signal } from '../decorators';
import { is } from '../utils';
import { Shape } from './Shape';
import { TxtLeaf } from './TxtLeaf';
let Txt = Txt_1 = class Txt extends Shape {
    /**
     * Create a bold text node.
     *
     * @remarks
     * This is a shortcut for
     * ```tsx
     * <Txt fontWeight={700} />
     * ```
     *
     * @param props - Additional text properties.
     */
    static b(props) {
        return new Txt_1({ ...props, fontWeight: 700 });
    }
    /**
     * Create an italic text node.
     *
     * @remarks
     * This is a shortcut for
     * ```tsx
     * <Txt fontStyle={'italic'} />
     * ```
     *
     * @param props - Additional text properties.
     */
    static i(props) {
        return new Txt_1({ ...props, fontStyle: 'italic' });
    }
    getText() {
        return this.innerText();
    }
    setText(value) {
        const children = this.children();
        let leaf = null;
        for (let i = 0; i < children.length; i++) {
            const child = children[i];
            if (leaf === null && child instanceof TxtLeaf) {
                leaf = child;
            }
            else {
                child.parent(null);
            }
        }
        if (leaf === null) {
            leaf = new TxtLeaf({ text: value });
            leaf.parent(this);
        }
        else {
            leaf.text(value);
        }
        this.setParsedChildren([leaf]);
    }
    setChildren(value) {
        if (this.children.context.raw() === value) {
            return;
        }
        if (typeof value === 'string') {
            this.text(value);
        }
        else {
            super.setChildren(value);
        }
    }
    *tweenText(value, time, timingFunction, interpolationFunction) {
        const children = this.children();
        if (children.length !== 1 || !(children[0] instanceof TxtLeaf)) {
            this.text.save();
        }
        const leaf = this.childAs(0);
        const oldText = leaf.text.context.raw();
        const oldSize = this.size.context.raw();
        leaf.text(value);
        const newSize = this.size();
        leaf.text(oldText ?? DEFAULT);
        if (this.height() === 0) {
            this.height(newSize.height);
        }
        yield* all(this.size(newSize, time, timingFunction), leaf.text(value, time, timingFunction, interpolationFunction));
        this.children.context.setter(value);
        this.size(oldSize);
    }
    getLayout() {
        return true;
    }
    constructor({ children, text, ...props }) {
        super(props);
        this.children(text ?? children);
    }
    innerText() {
        const children = this.childrenAs();
        let text = '';
        for (const child of children) {
            text += child.text();
        }
        return text;
    }
    parentTxt() {
        const parent = this.parent();
        return parent instanceof Txt_1 ? parent : null;
    }
    parseChildren(children) {
        const result = [];
        const array = Array.isArray(children) ? children : [children];
        for (const child of array) {
            if (child instanceof Txt_1 || child instanceof TxtLeaf) {
                result.push(child);
            }
            else if (typeof child === 'string') {
                result.push(new TxtLeaf({ text: child }));
            }
        }
        return result;
    }
    applyFlex() {
        super.applyFlex();
        this.element.style.display = this.findAncestor(is(Txt_1))
            ? 'inline'
            : 'block';
    }
    async draw(context) {
        await this.drawChildren(context);
    }
};
__decorate([
    initial(''),
    signal()
], Txt.prototype, "text", void 0);
__decorate([
    threadable()
], Txt.prototype, "tweenText", null);
__decorate([
    computed()
], Txt.prototype, "innerText", null);
__decorate([
    computed()
], Txt.prototype, "parentTxt", null);
Txt = Txt_1 = __decorate([
    nodeName('Txt')
], Txt);
export { Txt };
[
    'fill',
    'stroke',
    'lineWidth',
    'strokeFirst',
    'lineCap',
    'lineJoin',
    'lineDash',
    'lineDashOffset',
].forEach(prop => {
    Txt.prototype[`getDefault${capitalize(prop)}`] = function (initial) {
        return this.parentTxt()?.[prop]() ?? initial;
    };
});
