// @revideo/2d/lib/components/View2D.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var View2D_1;
import { PlaybackState, lazy } from '@revideo/core';
import { initial, signal } from '../decorators';
import { nodeName } from '../decorators/nodeName';
import { useScene2D } from '../scenes/useScene2D';
import { Rect } from './Rect';
let View2D = View2D_1 = class View2D extends Rect {
    constructor(props) {
        super({
            composite: true,
            fontFamily: 'Roboto',
            fontSize: 48,
            lineHeight: '120%',
            textWrap: false,
            fontStyle: 'normal',
            ...props,
        });
        this.view2D = this;
        View2D_1.shadowRoot.append(this.element);
        this.applyFlex();
    }
    dispose() {
        this.removeChildren();
        super.dispose();
    }
    async render(context) {
        this.computedSize();
        this.computedPosition();
        await super.render(context);
    }
    /**
     * Find a node by its key.
     *
     * @param key - The key of the node.
     */
    findKey(key) {
        return useScene2D().getNode(key) ?? null;
    }
    requestLayoutUpdate() {
        this.updateLayout();
    }
    requestFontUpdate() {
        this.applyFont();
    }
    view() {
        return this;
    }
};
__decorate([
    initial(PlaybackState.Paused),
    signal()
], View2D.prototype, "playbackState", void 0);
__decorate([
    initial(0),
    signal()
], View2D.prototype, "globalTime", void 0);
__decorate([
    initial(0),
    signal()
], View2D.prototype, "fps", void 0);
__decorate([
    signal()
], View2D.prototype, "assetHash", void 0);
__decorate([
    lazy(() => {
        const frameID = 'revideo-2d-frame';
        let frame = document.querySelector(`#${frameID}`);
        if (!frame) {
            frame = document.createElement('div');
            frame.id = frameID;
            frame.style.position = 'absolute';
            frame.style.pointerEvents = 'none';
            frame.style.top = '0';
            frame.style.left = '0';
            frame.style.opacity = '0';
            frame.style.overflow = 'hidden';
            document.body.prepend(frame);
        }
        return frame.shadowRoot ?? frame.attachShadow({ mode: 'open' });
    })
], View2D, "shadowRoot", void 0);
View2D = View2D_1 = __decorate([
    nodeName('View2D')
], View2D);
export { View2D };
